File: 2017-12-15 Vectors en Python/p3.py
# Genera totes les permutacions de n elements

from jutge import read


def genera(v, u, i):
    """Genera totes les permutacions havent fixat les primeres posicions en v i havent utilizat els valors en u."""
    n = len(v)
    if i == n:
        print(v)
    else:
        for j in range(n):
            if not u[j]:
                u[j] = True
                v[i] = j
                genera(v, u, i + 1)
                u[j] = False


def main():
    n = read(int)
    v = [None  for i in range(n)]
    u = [False for i in range(n)]
    genera(v, u, 0)


main()

File: 2017-12-05 Generació combinatòria/p3.cc
// Genera totes les permutacions dels n primers naturals.

#include <iostream>
#include <vector>
using namespace std;


void escriu(const vector<int>& p) {
    for (int i : p) cout << i;
    cout << endl;
}  


// Genera totes les permutacions del vector v deixant les 
// i primeres posicions fixades sabent que el vector u 
// descriu els elements que ja s'han usat a les primeres i posicions.
void genera(vector<int>& p, vector<bool>& u, int i) {
    int n = p.size();
    if (i == n) {
        escriu(p);
    } else {
        for (int j = 0; j < n; ++j) {
            if (not u[j]) {
                p[i] = j;
                u[j] = true;
                genera(p, u, i + 1);
                u[j] = false;
            }
        } 
    }
}


int main() {
    int n;
    cin >> n;

    vector<int> p(n);
    vector<bool> u(n, false);
    genera(p, u, 0);
}
